### DISCUSSION TITLE
Newt "Cloud/Processing Server" and APIs

### DISCUSSION THREAD

**Discussion by darianbjohnson** [Phambili-Tech/Newt_Display#39] [Ideas]
There's been a lot of discussion on the APIs that Newt uses, and how the information is gathered. I've taken a first pass at creating some documentation. I've also uploaded the source code...

Documentation (on the wiki)
https://github.com/Phambili-Tech/Newt_Display/wiki/MQTT-API-Summary

Source Code
https://github.com/Phambili-Tech/Newt_Display/blob/master/src/Newt_Cloud/Newt_APIs.js

**Comment by darianbjohnson** [comment]
@badslug @angusmci @tjfoth - I've started to document the APIs on the wiki, and I've uploaded the "Newt Cloud" (a nodeJS script that runs in AWS Lambda).

**Comment by badslug** [comment]
@darianbjohnson I'm curious why the system uses HTTP and MQTT. Could it be implemented with one or the other rather than both?

**Reply by darianbjohnson** [reply]
Great question/catch.

So, I do that because I'm still trying to find an optimal solution for an MQTT broker (I originally used AWS IOT, then planned to use Google IOT - before they went EOL; I currently use Shiftr.io). 

Anyway - I wanted the ability to dynamically change the MQTT server and connection information. So the initial connection is an https call to a hardcoded web-address (which I control and should ever change). The result includes the MQTT info

**Comment by tjfoth** [comment]
Darian,

Please don't take this as a strong recommendation, but I'm looking at
HiveMQ for my personal use.  I haven't tried it out yet.  I've seen some
pretty good reviews of it.  I don't know about the pricing however.  There
is a free option, but in the long term, that's not going to work for you.

Tom

---
Tom Foth
***@***.***
***@***.***

**Comment by darianbjohnson** [comment]
I've actually been thinking of just standing up my own MQTT server... you can standup cloud instances pretty cheaply in the cloud. I could start with one instance ($5 a month) and maybe later move to two instances (for HA purposes )

**Comment by badslug** [comment]
I guess my main question is "why MQTT"? If HTTP is already deployed on both ends, does MQTT add enough to be worth the overhead vs just having one communication channel. I'm trying to understand the benefits of MQTT for Newt.

**Reply by darianbjohnson** [reply]
Why MQTT - well, for a few reasons (some good, some kinda lazy)

1. I kinda always thought that I might switch from WiFi to another internet "protocol" - maybe Matter through a smart device/hub (HomeAssistant, Alexa, Google Home...), or ESPMesh/ESPnow... so having the code receive MQTT felt like a way to get ready for that
2. I didnt want the the http call to "block" the device if the user was using it. I was able to write a non-blocking call for wifi/mqtt connections... but couldn't get my head around writing a framework for non-blocking https GET calls
3. I didn't know how many "services" I would provide (weather, AQI, UV, Riddle, Quote, etc.), and I didn't want to write code for each GET call.

So, I kinda know what I'm doing now, so I think changing all the requests from MQTT to HTTPS GET would make sense. It would speed up the connection time (which means longer battery life) and it means that I wouldn't need to pay for an MQTT server....

So.... something to noodle on

**Reply by badslug** [reply]
Those all sound totally understandable for "why MQTT". I would advocate to moving to "just HTTPS" for reasons you probably already can guess:

1. Only one server to need to deploy and manage
2. Easier for users to deploy their own server (since they only need to deploy one)
3. Serverless https is easier to scale both in supporting just a single Newt and for an infinite number of Newts

Towards the last two points, it would be interesting to see if the code could be built so that if no database is configured the server side stuff works but essentially assumes a single newt with a single configuration (you could have more than one actual Newt but we could assume they are multiple Newts in the same house) or some other very low number of units scenarios that could be handled without a backing data store. Then if a database is configured, then we assume a multi-tenant set up. That would make it much simpler for individuals to fork and deploy the software without needing a database. Of course, I'm not sure if that is easy or not - it could be simpler to just have two servers - one for individuals and one more multi-tenant.

**Comment by tjfoth** [comment]
Yeah, I’m running an MQTT server at home and it is no big deal.  I just
like someone else being responsible for security.

On Sun, Sep 18, 2022 at 3:09 PM Darian Johnson ***@***.***>
wrote:

> I've actually been thinking of just standing up my own MQTT server... you
> can standup cloud instances pretty cheaply in the cloud. I could start with
> one instance ($5 a month) and maybe later move to two instances (for HA
> purposes )
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Phambili-Tech/Newt_Display/discussions/39#discussioncomment-3675790>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOIIO4CCALTBFLQUYXHDO3V65SINANCNFSM6AAAAAAQN7J6IA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>
-- 
—-
Tom Foth

***@***.***

**Comment by tjfoth** [comment]
My vote for MQTT is how nicely it integrates with home automation systems
and codeless automation systems like Node Red.

---
Tom Foth
***@***.***
***@***.***
